drivers/adc/gecko_adc.c

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) 2022 T-Mobile USA, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT silabs_adc_gecko
+
+#include <string.h>
+#include <zephyr/drivers/adc/gecko_adc/gecko_adc.h>
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+
+#include <em_cmu.h>
+#include <em_gpio.h>
+#include  "board.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(gecko_adc, CONFIG_ADC_LOG_LEVEL);
+
+K_SEM_DEFINE(adc_sem, 0, 1);
+
+static int get_reference_voltage(int reference)
+{
+	int ref;
+
+	switch (reference) {
+	case ADC_REF_INTERNAL:
+		ref = adcRef2V5;
+		break;
+	case ADC_REF_VDD_1:
+		ref = adcRefVDD;
+		break;
+	case ADC_REF_VDD_1_2:
+		ref = adcRef2V5;
+		break;
+	case ADC_REF_VDD_1_4:
+		ref = adcRef1V25;
+		break;
+	default:
+		ref = -1;
+		break;
+	}
+	return ref;
+}
+
+static int get_resolution(uint8_t resolution)
+{
+
+	switch (resolution) {
+	case 12:
+		return adcRes12Bit;
+	case 8:
+		return adcRes8Bit;
+	case 6:
+		return adcRes6Bit;
+	case 0:
+		return adcResOVS;
+	default:
+		LOG_ERR("ADC resolution value %d is not valid", resolution);
+		return -EINVAL;
+	}
+	return 0;
```

Comment:
Since there's a default return in the switch statement, this return is necessary.

Reply:
Removed